src/core/credentials/transport/xds/xds_credentials.cc

```diff
@@ -156,9 +197,26 @@ XdsCredentials::create_security_connector(
         }
       }
       tls_credentials_options->set_verify_server_cert(true);
+      if (XdsSniEnabled()) {
+        auto maybe_hostname = args->GetOwnedString(GRPC_ARG_ADDRESS_NAME);
+        if (xds_certificate_provider->auto_host_sni() &&
+            maybe_hostname.has_value()) {
+          tls_credentials_options->set_sni_override(maybe_hostname.value());
+        } else if (xds_certificate_provider->sni().length() > 0) {
+          tls_credentials_options->set_sni_override(
+              xds_certificate_provider->sni());
+        } else {
+          if (UseChannelAuthorityIfNoSNIApplicable()) {
+            tls_credentials_options->set_sni_override(std::nullopt);
+          } else {
+            tls_credentials_options->set_sni_override("");
+          }
+        }
+      }
       tls_credentials_options->set_certificate_verifier(
           MakeRefCounted<XdsCertificateVerifier>(
-              std::move(xds_certificate_provider)));
+              std::move(xds_certificate_provider),
+              tls_credentials_options->sni_override().value_or("")));
```

Comment:
It looks like if neither `auto_host_sni` nor `sni` are used, then we'll use the empty string for SAN validation.  Is that really what we want, or do we want to skip SAN validation entirely in that case?

It's not clear to me what the right behavior is from reading the gRFC.  I suggest checking how Java has implemented this, and trying to clarify the gRFC.

Reply:
The `auto_sni_san_validation` option controls whether that parameter is actually used. So, if neither `auto_host_sni` nor `sni` are used *and* `auto_sni_san_validation` is set, then yes, it will use the empty string for SAN validation. In my opinion, that is a sufficiently nonsensical configuration that it's OK for it to have strange behavior.